### DISCUSSION TITLE
Avalonia Equivalent of WPF

### DISCUSSION THREAD

**Discussion by BangCtrl** [AvaloniaUI/Avalonia#8627] [Q&A]
I'm trying to port something from WPF to Avalonia.
Some of the errors I've been having is:

Avalonia equivalent of the ContentPropertyAttribute. https://docs.microsoft.com/en-us/dotnet/api/system.windows.markup.contentpropertyattribute?view=windowsdesktop-6.0
For example, `[ContentProperty(nameof(EXAMPLE))]`

Avalonia equivalent of UIPropertyMetadata and FrameworkPropertyMetadata. https://docs.microsoft.com/en-us/dotnet/api/system.windows.uipropertymetadata?view=windowsdesktop-6.0
 For example, `public static readonly StyledProperty<string> TitleProperty = AvaloniaProperty.Register<Control, string>(nameof(Title), typeof(string), typeof(LayoutContent), new UIPropertyMetadata(null, OnTitlePropertyChanged, CoerceTitleValue));`

Avalonia equivalent of ImageSource. https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.imagesource?view=windowsdesktop-6.0
For example, `private ImageSource _iconSource = null;`

Avalonia equivalent of SetCurrentValue https://docs.microsoft.com/en-us/dotnet/api/system.windows.dependencyobject.setcurrentvalue?view=windowsdesktop-6.0

I'll be sure to put more errors I find as I port this app.

**Comment by maxkatz6** [comment]
> Avalonia equivalent of the ContentPropertyAttribute

`[Content]` attribute on the property itself. Or "Add(object)" method. Or `IAddChild` interface (master branch only, 11.0+).

> Avalonia equivalent of UIPropertyMetadata and FrameworkPropertyMetadata

[These props](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaProperty.cs#L235-L241). Internally it will create a metadata, which you can override per type if you need with OverrideMetadata/OverrideDefaultValue method.

If you need OnTitlePropertyChanged, use `Control.OnPropertyChanged` method override and check arguments. Or static `Property.Changed.AddClassHandler`.

> Avalonia equivalent of ImageSource

IImage interface. Avalonia Bitmap implements it. DrawingImage implements it too.

> Avalonia equivalent of SetCurrentValue

Not sure if you would need it with Avalonia. We have different binding value priorities that you can pass into SetValue method if you need set value with higher priority. SetValue also returns a IDisposable that will discard this value and return to a previous one on disposal. 

For more information, see our documentation https://docs.avaloniaui.net/ . It also has tips for WPF developer.
And you can find a lot of information in the github issues.

**Reply by BangCtrl** [reply]
@maxkatz6, thank you for your answer. However, content attribute does not work. As for FrameworkPropertyMetadata and UIPropertyMetadata, after putting `public static readonly AvaloniaProperty ContentIdProperty = AvaloniaProperty.Register(nameof(ContentId), typeof(string), typeof(LayoutContent), new UIPropertyMetadata(null, OnContentIdPropertyChanged));` I am getting compiler error CS0411 https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0411?f1url=%3FappId%3Droslyn%26k%3Dk(CS0411).

**Reply by maxkatz6** [reply]
Check documentation https://docs.avaloniaui.net/docs/authoring-controls/defining-properties
You are using WPF syntrax
